src/test/java/net/minecraftforge/debug/world/MultiNoiseBiomeSourceLoadingEventTest.java

```diff
@@ -0,0 +1,274 @@
+    
+    public static final ResourceLocation PRESET_TEST = new ResourceLocation(MODID, "preset_test");
+    public static final ResourceLocation NAMED_TEST = new ResourceLocation(MODID, "named_test");
+    public static final ResourceLocation NAMELESS_TEST = new ResourceLocation(MODID, "nameless_test");
+    
+    public MultiNoiseBiomeSourceLoadingEventTest()
+    {
+        IEventBus modBus = FMLJavaModLoadingContext.get().getModEventBus();
+        IEventBus forgeBus = MinecraftForge.EVENT_BUS;
+
+        modBus.addListener(this::onGatherData);
+        forgeBus.addListener(this::onLoadBiomeSource);
+        forgeBus.addListener(this::onRegisterCommands);
+    }
+    
+    private void onGatherData(GatherDataEvent event)
+    {
+        if (event.includeServer())
+        {
+            this.onGatherServerData(event);
+        }
+    }
+    
+    private void onGatherServerData(GatherDataEvent event)
+    {
+        // need dynamic registries to do worldgen datagen
+        RegistryHolder registries = RegistryAccess.builtin();
+        RegistryWriteOps<JsonElement> writeOps = RegistryWriteOps.create(JsonOps.INSTANCE, registries);
+        
+        // the objects we reference must have been registered to the registryaccess
+        // creating a registry access deep-copies most of the BuiltinRegistries objects,
+        // so we must get vanilla objects from the registryaccess instead of getting static references
+        Registry<DimensionType> dimensionTypes = registries.registryOrThrow(Registry.DIMENSION_TYPE_REGISTRY);
+        Registry<NoiseGeneratorSettings> noiseSettings = registries.registryOrThrow(Registry.NOISE_GENERATOR_SETTINGS_REGISTRY);
+        Registry<NoiseParameters> noiseParameters = registries.registryOrThrow(Registry.NOISE_REGISTRY);
+        Registry<Biome> biomes = registries.registryOrThrow(Registry.BIOME_REGISTRY);
+        
+        DimensionType overworldDimensionType = dimensionTypes.get(DimensionType.OVERWORLD_LOCATION);
+        NoiseGeneratorSettings overworldNoiseSettings = noiseSettings.get(NoiseGeneratorSettings.OVERWORLD);
+        Biome desert = biomes.get(Biomes.DESERT);
+        
+        // make three dimensions for testing
+        // one with a preset biome source, one with a fully custom biome source and a dimension name, one with a fully custom source without name
+        
+        // preset biome source dimension
+        BiomeSource presetSource = MultiNoiseBiomeSource.Preset.OVERWORLD.biomeSource(biomes);
+        ChunkGenerator presetGenerator = new NoiseBasedChunkGenerator(noiseParameters, presetSource, 0, () -> overworldNoiseSettings);
+        LevelStem presetDimension = new LevelStem(() -> overworldDimensionType, presetGenerator);
+        
+        // some sample points used for the latter dimensions
+        List<Pair<ParameterPoint, Supplier<Biome>>> parameterPoints = new ArrayList<>();
+        parameterPoints.add(Pair.of(new ParameterPoint(
```

Comment:
A minor quibble: this (and the `onLoadBiomeSource` listener below) should use the newly-added `addParameter` convenience.